src/prefect/client/base.py

```diff
@@ -154,27 +155,74 @@ class PrefectHttpxClient(httpx.AsyncClient):
-            await super().send(*args, **kwargs)
-        )
-        while (
-            response.status_code
-            in {status.HTTP_429_TOO_MANY_REQUESTS, status.HTTP_503_SERVICE_UNAVAILABLE}
-            and retry_count < self.RETRY_MAX
-        ):
-            retry_count += 1
-
-            # Respect the "Retry-After" header, falling back to an exponential back-off
-            retry_after = response.headers.get("Retry-After")
-            if retry_after:
-                retry_seconds = float(retry_after)
+    async def _send_with_retry(
+        self,
+        request: Callable,
+        retry_codes: Set[int] = set(),
+        retry_exceptions: Tuple[Exception, ...] = tuple(),
+    ):
+        """
+        Send a request and retry it if it fails.
+
+        Sends the provided request and retries it up to self.RETRY_MAX times if
+        the request either raises an exception listed in `retry_exceptions` or receives
+        a response with a status code listed in `retry_codes`.
+
+        Retries will be delayed based on either the retry header (preferred) or
+        exponential backoff if a retry header is not provided.
+        """
+        try_count = 0
+        response = None
+        retry = True
+
+        while retry:
+            try_count += 1
+            retry_seconds = None
+
+            try:
+                response = await request()
+            except retry_exceptions as exc:
+                if try_count > self.RETRY_MAX:
+                    retry = False
+                    raise
+                else:
+                    continue
+            except BaseException as exc:
+                retry = False
+                raise
+
             else:
-                retry_seconds = 2**retry_count
+                # we got a good response
+                if response.status_code not in retry_codes:
+                    retry = False
+                # respect retry headers
+                else:
+                    retry_after = response.headers.get("Retry-After")
+                    if retry_after:
+                        retry_seconds = float(retry_after)
+
+            finally:
+                if retry:
+                    if retry_seconds is None:
+                        retry_seconds = 2**try_count
+                    await anyio.sleep(retry_seconds)
+
+                    if try_count > self.RETRY_MAX:
+                        retry = False
+
+        return response
 
-            await anyio.sleep(retry_seconds)
-            response = await super().send(*args, **kwargs)
+    async def send(self, *args, **kwargs) -> Response:
+        api_request = partial(super().send, *args, **kwargs)
+
+        response = await self._send_with_retry(
+            request=api_request,
+            retry_codes={
+                status.HTTP_429_TOO_MANY_REQUESTS,
+                status.HTTP_503_SERVICE_UNAVAILABLE,
+            },
+            retry_exceptions=(httpx.RemoteProtocolError, httpx.ReadError),
```

Comment:
Hi.

Just saw your PR and tested it. In our case it didn't help until I updated the initial set of exceptions to the following:

```python
retry_exceptions=(httpx.RemoteProtocolError, httpx.ReadError, httpx.PoolTimeout, httpx.ReadTimeout)
```
With this update the workflow runs without any problems, but without it we still got crashing executions.

Reply:
Awesome - thanks a ton for giving it a shot and following up. I'll get those errors added. 

@madkinsz 
Something that I'm wondering about is if we would possibly want to consider making these exceptions configurable by an environment variable or something similar so that users can adjust on the fly?

Reply:
There have been requests for that, but let's consider it separately.

Reply:
@peytonrunyan let's get this finished today and include the local protocol error in https://github.com/PrefectHQ/prefect/issues/7442